aten/src/THC/THCReduce.cuh

```diff
@@ -63,58 +63,68 @@ kernelReduceNoncontigDim_shared(TensorInfo<T, IndexType> out,
+        const AccT val1 = scalar_cast<AccT>(in.data[inOffset + (i + blockDim.y) * reductionStride]);
+        load_reg[1] = modifyOp(val1);
+        const AccT val2 = scalar_cast<AccT>(in.data[inOffset + (i + blockDim.y * 2) * reductionStride]);
+        load_reg[2] = modifyOp(val2);
+        const AccT val3 = scalar_cast<AccT>(in.data[inOffset + (i + blockDim.y * 3) * reductionStride]);
+        load_reg[3] = modifyOp(val3);
         local_reg = reduceOp(local_reg, load_reg[0]);
         local_reg = reduceOp(local_reg, load_reg[1]);
         local_reg = reduceOp(local_reg, load_reg[2]);
         local_reg = reduceOp(local_reg, load_reg[3]);
-      }else if(i + blockDim.y * 2 < reductionSize){
-        load_reg[0] = modifyOp(in.data[inOffset + (i + blockDim.y * 0) * reductionStride]);
-        load_reg[1] = modifyOp(in.data[inOffset + (i + blockDim.y * 1) * reductionStride]);
-        load_reg[2] = modifyOp(in.data[inOffset + (i + blockDim.y * 2) * reductionStride]);
+      } else if (i + blockDim.y * 2 < reductionSize) {
+        const AccT val0 = scalar_cast<AccT>(in.data[inOffset + i * reductionStride]);
+        load_reg[0] = modifyOp(val0);
+        const AccT val1 = scalar_cast<AccT>(in.data[inOffset + (i + blockDim.y) * reductionStride]);
+        load_reg[1] = modifyOp(val1);
+        const AccT val2 = scalar_cast<AccT>(in.data[inOffset + (i + blockDim.y * 2) * reductionStride]);
+        load_reg[2] = modifyOp(val2);
         local_reg = reduceOp(local_reg, load_reg[0]);
         local_reg = reduceOp(local_reg, load_reg[1]);
         local_reg = reduceOp(local_reg, load_reg[2]);
-      }else if( (i + blockDim.y) < reductionSize){
-        load_reg[0] = modifyOp(in.data[inOffset + (i + blockDim.y * 0) * reductionStride]);
-        load_reg[1] = modifyOp(in.data[inOffset + (i + blockDim.y * 1) * reductionStride]);
+      } else if (i + blockDim.y < reductionSize) {
+        const AccT val0 = scalar_cast<AccT>(in.data[inOffset + i * reductionStride]);
+        load_reg[0] = modifyOp(val0);
+        const AccT val1 = scalar_cast<AccT>(in.data[inOffset + (i + blockDim.y) * reductionStride]);
+        load_reg[1] = modifyOp(val1);
         local_reg = reduceOp(local_reg, load_reg[0]);
         local_reg = reduceOp(local_reg, load_reg[1]);
-      }else if(i + blockDim.y * 0 < reductionSize){
-        local_reg = reduceOp(local_reg, modifyOp(in.data[inOffset + i * reductionStride]));
+      } else if (i < reductionSize) {
+        const AccT val0 = scalar_cast<AccT>(in.data[inOffset + i * reductionStride]);
+        local_reg = reduceOp(local_reg, modifyOp(val0));
       }
     }
 
     *shmem = local_reg;
     int dimy = blockDim.y;
-    while(dimy > 1){
+    while (dimy > 1) {
       __syncthreads();
-      if( threadIdx.y == 0 && (dimy%2 != 0) ){
-        *shmem = reduceAccOp(*shmem, *(shmem + (dimy-1) * blockDim.x) );
+      if (threadIdx.y == 0 && (dimy % 2 != 0) ) {
+        *shmem = reduceOp(*shmem, *(shmem + (dimy - 1) * blockDim.x));
```

Comment:
So... shmem is an AccT, right? So it's weird to see the diff remove Acc so it's `reduceOp` now...

Reply:
Hmm... 

So before this PR there was reduceOp and reduceAccOp. However, this PR simplifies things so there is now only a single reduceOp that takes in two AccT and returns one. This is because the new pattern is:

1) Acquire the data.
2) Convert it to the accumulate type.
3) Perform operations (like modify and reduce and finalize) on it
4) Convert the out value to the "real" type and write it out

If you're concerned about some implicit conversions here we have some added safety because half types cannot be implicitly converted to other types (like float), so unless we're very careful with our (explicit) conversions we will not be able to compile. 

Does that make sense?

Reply:
Yes, I think my suggestion here is actually quite simple: call it `reduceAccOp` instead of `reduceOp`, because it takes `AccT` :)

Reply:
Oh I see. We could change the name, but I think that'd be verbose. The advantage is that it would remind you that you were operating on the accum type (good), but that's the standard and ModifyOp and FinalizeOp both also work on the accum type. So then we'd have to change them to be ModifyAccOp and FinalizeAccOp, and maybe change from type T to be type AccT everywhere we expected the accum type vs. any type, etc.

I'll do it if you like but if we're worried about clarifying what the operators are working on then a comment may be better. I put a brief on in THCTensorMathReduce.cuh, but could elaborate on it and/or add comments elsewhere, too.

Reply:
Speaking of implicit conversions, right now no implicit conversions and no half math hinges on a line in aten CMakeLists https://github.com/pytorch/pytorch/blob/f7c9580fd3699f54f48fbf5d0a182265e752ae18/aten/CMakeLists.txt#L138. With all the refactoring and c2 merge going on, it seems very easy to lose this line, or to have some parts of the code that don't use aten's CMakeLists be compiled with implicit conversions. @ezyang, is there a way to make disabling implicit conversions more robust?

Reply:
@mruberry OK, you've convinced me :)

@ngimel Here's an idea. In C++, you can turn off implicit conversion by making a function templated. So you could do something like this: https://godbolt.org/g/zz3EcU